In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error
from seaborn import regplot
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [2]:
def SelectFeature(feature_candidates,features_used,targets,df):
    N = len(feature_candidates)
    R2 = np.zeros(N)
    for k in range(N):   
        features_current = features_used.copy()
        features_current.append(feature_candidates[k])
        features = df[features_current]
        results = cross_validate(lr,features,targets,n_jobs=-1)
        R2[k] = results['test_score'].mean()
        
    R2_max = R2.max()
    feature_selected = feature_candidates[R2.argmax()] 
    return (feature_selected,R2_max)

In [3]:
df = pd.read_csv("./bacteria.csv")
print(df.shape)
df.head()

(100, 4)


temperature  humidity surface   CFU
0         23.2       0.7   rough  33.0
1         21.7       0.7   rough  15.0
2         15.1       0.5  smooth  28.0
3         30.8       0.2   rough  17.0
4         11.2       0.5   rough  14.0

In [4]:
features = df.drop('CFU',axis=1)
features.head(3)

temperature  humidity surface
0         23.2       0.7   rough
1         21.7       0.7   rough
2         15.1       0.5  smooth

In [5]:
features = pd.get_dummies(features)

In [6]:
features.head(3)

temperature  humidity  surface_rough  surface_smooth
0         23.2       0.7              1               0
1         21.7       0.7              1               0
2         15.1       0.5              0               1

In [7]:
target = df.CFU

In [8]:
target.head(3)

0    33.0
1    15.0
2    28.0
Name: CFU, dtype: float64

In [9]:
#features = (features - features.mean()) / features.std()
lr = LinearRegression()
poly = PolynomialFeatures(2)
features_engineered = poly.fit_transform(features)
cols = poly.get_feature_names(features.columns)
features_engineered = pd.DataFrame(features_engineered,columns=cols)
features_engineered.head(3)
print('features shape =',features.shape)
print('features (engineered) shape =',features_engineered.shape)

features shape = (100, 4)
features (engineered) shape = (100, 15)


In [10]:
ix = (features_engineered.std() == 0)
drop_cols = features_engineered.columns[ix]
features_engineered = features_engineered.drop(drop_cols,axis=1)
print('features (engineered) shape =',features_engineered.shape)

features (engineered) shape = (100, 13)


In [11]:
features_engineered = (features_engineered - features_engineered.mean())/features_engineered.std()
features_engineered.describe()

temperature      humidity  surface_rough  surface_smooth  \
count  1.000000e+02  1.000000e+02   1.000000e+02    1.000000e+02   
mean   9.325873e-17  1.534328e-15   1.287859e-16   -1.287859e-16   
std    1.000000e+00  1.000000e+00   1.000000e+00    1.000000e+00   
min   -1.617928e+00 -1.901657e+00  -1.015090e+00   -9.752828e-01   
25%   -8.739583e-01 -5.181764e-01  -1.015090e+00   -9.752828e-01   
50%   -4.759467e-02  1.106784e-01   9.752828e-01   -9.752828e-01   
75%    9.629438e-01  1.116846e+00   9.752828e-01    1.015090e+00   
max    1.755380e+00  1.619930e+00   9.752828e-01    1.015090e+00   

       temperature^2  temperature humidity  temperature surface_rough  \
count   1.000000e+02          1.000000e+02               1.000000e+02   
mean   -5.773160e-17          7.804868e-16              -3.663736e-17   
std     1.000000e+00          1.000000e+00               1.000000e+00   
min    -1.180563e+00         -1.442449e+00              -8.580065e-01   
25%    -8.897355e-01         -7.547051e-01              -8.580065e-01   
50%    -2.709141e-01         -2.964492e-01              -3.822388e-01   
75%     9.092876e-01          6.948800e-01               7.859064e-01   
max     2.160652e+00          2.473949e+00               2.081697e+00   

       temperature surface_smooth    humidity^2  humidity surface_rough  \
count                1.000000e+02  1.000000e+02            1.000000e+02   
mean                -8.881784e-18 -4.440892e-18           -1.776357e-17   
std                  1.000000e+00  1.000000e+00            1.000000e+00   
min                 -7.971138e-01 -1.338537e+00           -8.945165e-01   
25%                 -7.971138e-01 -6.500425e-01           -8.945165e-01   
50%                 -7.971138e-01 -9.145281e-02           -3.534783e-01   
75%                  7.820739e-01  1.155631e+00            9.089442e-01   
max                  2.195823e+00  1.935058e+00            1.991021e+00   

       humidity surface_smooth  surface_rough^2  surface_smooth^2  
count             1.000000e+02     1.000000e+02      1.000000e+02  
mean              1.332268e-17     1.287859e-16     -1.287859e-16  
std               1.000000e+00     1.000000e+00      1.000000e+00  
min              -8.241727e-01    -1.015090e+00     -9.752828e-01  
25%              -8.241727e-01    -1.015090e+00     -9.752828e-01  
50%              -8.241727e-01     9.752828e-01     -9.752828e-01  
75%               6.091712e-01     9.752828e-01      1.015090e+00  
max               2.042515e+00     9.752828e-01      1.015090e+00

In [12]:
lr = LinearRegression()
feature_candidates = list(features_engineered.columns)
features_used = []
R2 = []

while (len(feature_candidates) > 0):
    print('candidates left:',len(feature_candidates),end='')
    (feature_selected,R2_max) = SelectFeature(feature_candidates,features_used,target,features_engineered)
    features_used.append(feature_selected)
    feature_candidates.remove(feature_selected)
    R2.append(R2_max)
    print('  selected:',feature_selected)
    
results = pd.DataFrame()
results['features'] = features_used
results['test R-squared'] = R2

print()
print(results)
ix = results['test R-squared'].idxmax()
print()
print('use features 0 to',ix,end='')
R2_max = results['test R-squared'].max()
print('    maximum test R-squared =',R2_max.round(3))

candidates left: 13  selected: temperature surface_smooth
candidates left: 12  selected: temperature surface_rough
candidates left: 11  selected: surface_rough
candidates left: 10  selected: surface_smooth
candidates left: 9  selected: temperature
candidates left: 8  selected: surface_rough^2
candidates left: 7  selected: surface_smooth^2
candidates left: 6  selected: temperature^2
candidates left: 5  selected: temperature humidity
candidates left: 4  selected: humidity^2
candidates left: 3  selected: humidity surface_rough
candidates left: 2  selected: humidity surface_smooth
candidates left: 1  selected: humidity

                      features  test R-squared
0   temperature surface_smooth        0.438293
1    temperature surface_rough        0.521104
2                surface_rough        0.618322
3               surface_smooth        0.618322
4                  temperature        0.618322
5              surface_rough^2        0.618322
6             surface_smooth^2        0.618322


In [13]:
features = (features - features.mean()) / features.std()
lr.fit(features,target)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [14]:
R2 = lr.score(features, target)
print('R-squared = ', round(R2,3))

R-squared =  0.671


### The features engineered does not increase the R-squared value. 0.671 > 0.618. So, I will use the normal features, rather than engineered ones

In [15]:
coef = pd.DataFrame()
coef['feature'] = features.columns
coef['coef'] = lr.coef_.round(2)
coef.sort_values('coef',ascending=False)

feature  coef
0     temperature  9.99
3  surface_smooth  4.63
1        humidity  0.62
2   surface_rough -4.63

### From here, we see for every one CFU, temperature increases by almost 10 units, and humidity goes up 0.62 units. As you increase CFU, the likelihood that the surface is smooth increases, and likewise the likelihood that the surface is rough decreases

In [16]:
df = pd.read_csv("./bacteria_predict.csv")
df.head(3)

temperature  humidity surface  CFU
0         18.2       0.5  smooth    0
1          7.3       0.4  smooth    0
2          6.0       0.2  smooth    0

In [17]:
features = df.drop('CFU',axis=1)
features = pd.get_dummies(features)
features = (features - features.mean()) / features.std()

In [18]:
predictions = lr.predict(features)

In [19]:
df['CFU'] = predictions
df

temperature  humidity surface        CFU
0          18.2       0.5  smooth  38.170834
1           7.3       0.4  smooth  26.683532
2           6.0       0.2  smooth  24.739695
3          26.5       0.1   rough  27.033898
4          30.6       0.2  smooth  49.976615
5          16.9       0.4  smooth  36.532086
6          27.0       0.1   rough  27.546844
7          31.1       0.5  smooth  51.404829
8           8.9       0.6   rough  10.503663
9           9.0       0.8   rough  11.216431
10         36.6       0.4  smooth  56.742140
11         21.9       0.8   rough  24.450426
12          8.7       0.4   rough   9.688306
13         12.5       0.5  smooth  32.323255
14         31.5       0.2   rough  32.468443
15         37.2       0.5  smooth  57.662764
16         24.2       0.3   rough  25.284528
17         28.3       0.2  smooth  47.617065
18         10.6       0.7  smooth  30.984241
19         24.8       0.1   rough  25.289884
20          8.3       0.8   rough  10.498308
21         15.7       0.6   rough  17.479723
22         12.2       0.4   rough  13.278925
23         28.3       0.6  smooth  48.837423
24         36.6       0.1  smooth  55.826872
25         27.7       0.8   rough  30.400594
26         13.8       0.3   rough  14.615261
27         20.6       0.6  smooth  40.938062
28         24.0       0.6  smooth  44.426092
29         31.4       0.1   rough  32.060765
..          ...       ...     ...        ...
70         36.1       0.2  smooth  55.619016
71         25.6       0.3  smooth  45.152249
72         14.2       0.6  smooth  34.372360
73         25.7       0.8  smooth  46.780286
74         22.4       0.5  smooth  42.479577
75          5.6       0.4  smooth  24.939518
76         10.5       0.5   rough  11.840000
77         30.2       0.4   rough  31.744964
78         38.4       0.5  smooth  58.893833
79         21.3       0.3  smooth  40.740918
80         25.1       0.3  smooth  44.639304
81         17.4       0.4  smooth  37.045032
82         27.7       0.5  smooth  47.916799
83         39.1       0.4  smooth  59.306868
84         13.2       0.6  smooth  33.346469
85         36.9       0.7   rough  39.533702
86         32.3       0.5  smooth  52.635898
87         29.8       0.7  smooth  50.681349
88         24.2       0.4  smooth  44.021091
89         14.1       0.6  smooth  34.269770
90         20.7       0.3  smooth  40.125383
91         21.3       0.6  smooth  41.656186
92         21.6       0.1   rough  22.007032
93         28.7       0.5   rough  30.511217
94         30.0       0.6   rough  32.149965
95         19.6       0.2  smooth  38.691813
96          8.8       0.4   rough   9.790895
97         27.0       0.3  smooth  46.588497
98         10.3       0.4   rough  11.329732
99         39.4       0.6   rough  41.793340

[100 rows x 4 columns]

In [21]:
df.to_csv("bactera_rileyma.csv",index=False)
df.head(3)

temperature  humidity surface        CFU
0         18.2       0.5  smooth  38.170834
1          7.3       0.4  smooth  26.683532
2          6.0       0.2  smooth  24.739695